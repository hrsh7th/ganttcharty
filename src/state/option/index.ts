import startOfDay from 'date-fns/start_of_day';
import * as State from '../';
import { Column } from '../../component/ui-kit/grid';

export const DAY = 24 * 60 * 60 * 1000;

export const WEEK = DAY * 7;

export type Option = {
  headerWidth: number;
  rowHeight: number;
  columnWidth: number;
  scale: Scale;
  axisHeight: number;
  indentWidth: number;
  baseTime: Date;
  dayLabel: string[];
  columns: Column<State.Task.TaskNode>[];
};

export type Scale = 'day' | 'week';

/**
 * create scale time.
 */
export const scaleTime = (scale: Scale) => ({
  day: DAY,
  week: WEEK
}[scale]);

export const defaults = (option: Partial<Option> = {}) => {
  option.headerWidth = 240;
  option.rowHeight = 24;
  option.columnWidth = 18;
  option.scale = 'day';
  option.axisHeight = 32;
  option.indentWidth = 12;
  option.baseTime = startOfDay(new Date(Date.now() - scaleTime(option.scale) * 7));
  option.dayLabel = [
    '日',
    '月',
    '火',
    '水',
    '木',
    '金',
    '土'
  ];
  option.columns = [
    {
      key: 'name',
      name: 'タスク',
      width: 200

    },
    {
      key: 'startedAt',
      name: '開始日',
      width: 120

    },
    {
      key: 'finishedAt',
      name: '終了日',
      width: 120

    }
  ];
  return option as Option;
};

