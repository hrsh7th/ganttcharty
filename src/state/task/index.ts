import memoize from 'memoize-one';
import startOfDay from 'date-fns/start_of_day';

/**
 * task types.
 */
export type TaskId = string;
export type Task = {
  id: TaskId;
  startedAt: Date;
  finishedAt: Date;
  name: string;
  description: string;
  parentId?: TaskId;
  collapsed?: boolean;
};

/**
 * task tree types.
 */
export type TaskNode = Task & {
  parent?: Task;
  children: TaskNode[];
  depth: number;
};

/**
 * create tasks.
 */
export const tasks = memoize((tasks: Task[]) => {
  return (function traverse(
    tasks: Task[],
    parentId?: TaskId,
    depth: number = 0
  ): TaskNode[] {
    const parent = getTask(tasks, parentId);
    return tasks
      .filter(task => task.parentId === parentId)
      .reduce((nodes, task) => {
        const children = !task.collapsed ? traverse(tasks, task.id, depth + 1) : [];
        return nodes
          .concat([{
            ...task,
            parent,
            children,
            depth
          }])
          .concat(children);
      }, [] as TaskNode[]);
  })(tasks);
});

/**
 * get task by id.
 */
export const getTask = (tasks: Task[], taskId?: TaskId) => {
  return tasks.find(task => String(task.id) === String(taskId));
};

/**
 * get children by id.
 */
export const getChildren = (tasks: Task[], taskId?: TaskId, checkExpanded: boolean = true) => {
  const target = getTask(tasks, taskId)!;
  if (target && (target.collapsed && checkExpanded)) {
    return [];
  }
  return tasks.filter(task => task.parentId === taskId);
};

/**
 * get siblings by id.
 */
export const getSiblings = (tasks: Task[], taskId: TaskId) => {
  const target = getTask(tasks, taskId)!;
  const parent = getTask(tasks, target.parentId);
  return getChildren(tasks, parent ? parent.id : undefined);
};

/**
 * get siblings prev by id.
 */
export const getSiblingPrev = (tasks: Task[], taskId: TaskId): Task | null => {
  const target = getTask(tasks, taskId)!;
  const siblings = getSiblings(tasks, target.id);
  const index = siblings.indexOf(target);
  if (siblings[index - 1]) {
    return siblings[index - 1];
  }
  return null;
};

/**
 * get siblings next by id.
 */
export const getSiblingNext = (tasks: Task[], taskId: TaskId): Task | null => {
  const target = getTask(tasks, taskId)!;
  const siblings = getSiblings(tasks, target.id);
  const index = siblings.indexOf(target);
  if (siblings[index + 1]) {
    return siblings[index + 1];
  }
  return null;
};

/**
 * useful getting prev task by id.
 */
export const getPrev = (tasks: Task[], taskId: TaskId): Task | null => {
  const prev = getSiblingPrev(tasks, taskId);
  if (prev) {
    const children = getChildren(tasks, prev.id);
    if (children.length) {
      return children[children.length - 1];
    }
    return prev;
  }

  const target = getTask(tasks, taskId)!;
  const parent = getTask(tasks, target.parentId);
  if (parent) {
    return parent;
  }
  return null;
};

/**
 * useful getting next task by id.
 */
export const getNext = (tasks: Task[], taskId: TaskId): Task | null => {
  const children = getChildren(tasks, taskId);
  if (!!children.length) {
    return children[0];
  }

  const next = getSiblingNext(tasks, taskId);
  if (next) {
    return next;
  }

  const target = getTask(tasks, taskId)!;
  const parent = getTask(tasks, target.parentId);
  if (parent) {
    const next = getSiblingNext(tasks, parent.id);
    if (next) {
      return next;
    }
  }
  return null;
};

export const defaults = (tasks: Task[]) => {
  return tasks.sort((a, b) => a.startedAt.getTime() - b.startedAt.getTime()).map(task => {
    task.startedAt = startOfDay(task.startedAt);
    task.finishedAt = startOfDay(task.finishedAt);
    return task;
  });
};

