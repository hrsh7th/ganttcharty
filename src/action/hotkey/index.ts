import * as Action from '../';
import * as State from '../../state';
import throttle from 'lodash/throttle';

export const keyMap = {
  'selected up': ['k', 'up'],
  'selected down': ['j', 'down'],
  'move left': ['h', 'left'],
  'move right': ['l', 'right'],
  'add task': ['enter'],
  'delete task': ['backspace'],
  'expand task': ['shift+right'],
  'collapse task': ['shift+left']
};

export const handlers = {
  'selected up': () => {
    Action.Task.selectPrevTask();
  },
  'selected down': () => {
    Action.Task.selectNextTask();
  },
  'move left': throttle(() => {
    const state = State.get()!;
    Action.Task.moveSelectedTask(-State.Option.scaleTime(state.option.scale));
  }, 60),
  'move right': throttle(() => {
    const state = State.get()!;
    Action.Task.moveSelectedTask(State.Option.scaleTime(state.option.scale));
  }, 60),
  'add task': (event?: KeyboardEvent) => {
    event!.preventDefault();
    Action.Task.add();
  },
  'delete task': () => {
    Action.Task.deleteSelectedTask();
  },
  'expand task': () => {
    const state = State.get()!;
    if (state.ui.selectedTaskId) {
      Action.Task.expand(state.ui.selectedTaskId);
    }
  },
  'collapse task': () => {
    const state = State.get()!;
    if (state.ui.selectedTaskId) {
      Action.Task.collapse(state.ui.selectedTaskId);
    }
  }
};

