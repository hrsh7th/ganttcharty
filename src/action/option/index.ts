import * as State from '../../state';

export const zoomup = () => {
  State.update(state => {
    state.option.columnWidth += 3;
  });
};

export const zoomdown = () => {
  State.update(state => {
    state.option.columnWidth -= 3;
  });
};

export const changeHeaderWidth = (width: number) => {
  State.update(state => {
    state.option.headerWidth = width;
  });
};
