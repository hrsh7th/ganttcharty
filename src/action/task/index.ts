import * as uuid from 'uuid';
import * as State from '../../state';

export const updateTask = (taskId: State.Task.TaskId, attrs: Partial<State.Task.Task>) => {
  State.update(state => {
    const task = State.Task.getTask(state.tasks, taskId);
    Object.keys(attrs).forEach(key => {
      // @ts-ignore
      task[key] = attrs[key];
    });
  })
};

export const moveSelectedTask = (adder: number) => {
  State.update(state => {
    if (!state.ui.selectedTaskId) return;

    const target = State.Task.getTask(state.tasks, state.ui.selectedTaskId)!;
    target.startedAt = new Date(target.startedAt.getTime() + adder);
    target.finishedAt = new Date(target.finishedAt.getTime() + adder);
  });
};

export const deleteSelectedTask = () => {
  State.update(state => {
    if (!state.ui.selectedTaskId) return;

    // remove selected task.
    const target = State.Task.getTask(state.tasks, state.ui.selectedTaskId)!;

    // remove children.
    const children = State.Task.getChildren(state.tasks, target.id);
    children.forEach(child => state.tasks.splice(state.tasks.indexOf(child), 1));

    // memory target to move.
    const prev = State.Task.getPrev(state.tasks, state.ui.selectedTaskId);
    const next = State.Task.getNext(state.tasks, state.ui.selectedTaskId);

    // remove target task.
    state.tasks.splice(state.tasks.indexOf(target), 1);

    // move to target.
    if (next) {
      state.ui.selectedTaskId = next.id;
    } else if (prev) {
      state.ui.selectedTaskId = prev.id;
    }

    // clear selecting if remove last task.
    if (state.tasks.length === 0) {
      state.ui.selectedTaskId = undefined;
    }
  });
};

export const selectNextTask = () => {
  State.update(state => {
    if (!state.ui.selectedTaskId) return;

    const next = State.Task.getNext(state.tasks, state.ui.selectedTaskId);
    if (next) {
      state.ui.selectedTaskId = next.id;
    }
  });
};

export const selectPrevTask = () => {
  State.update(state => {
    if (!state.ui.selectedTaskId) return;

    const prev = State.Task.getPrev(state.tasks, state.ui.selectedTaskId);
    if (prev) {
      state.ui.selectedTaskId = prev.id;
    }
  });
};

export const expand = (taskId: State.Task.TaskId) => {
  State.update(state => {
    const selected = State.Task.getTask(state.tasks, taskId)!;
    const children = State.Task.getChildren(state.tasks, selected.id, false);
    if (selected && children.length && selected.collapsed) {
      selected.collapsed = false;
    }
  });
};

export const collapse = (taskId: State.Task.TaskId) => {
  State.update(state => {
    const selected = State.Task.getTask(state.tasks, taskId)!;
    const children = State.Task.getChildren(state.tasks, selected.id, false);
    if (selected && children.length &&  !selected.collapsed) {
      selected.collapsed = true;
    }
  });
};

export const add = () => {
  State.update(state => {
    if (!state.ui.selectedTaskId) return;

    // create newTask and extends selected task.
    const selected = State.Task.getTask(state.tasks, state.ui.selectedTaskId)!;
    const newTask: State.Task.Task = {
      id: uuid.v4(),
      name: 'new task',
      description: '',
      startedAt: new Date(selected.startedAt.getTime()),
      finishedAt: new Date(selected.finishedAt.getTime()),
      parentId: selected.parentId
    };

    // insert newTask to next to selected task.
    state.tasks.splice(
      state.tasks.indexOf(selected) + 1,
      0,
      newTask
    );

    // select newTask.
    state.ui.selectedTaskId = newTask.id;
  });
};

