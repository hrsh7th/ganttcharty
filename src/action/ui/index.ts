import * as State from '../../state';

export const updateCurrentTime = (diffX: number) => {
  State.update(state => {
    const diffTime = State.Option.scaleTime(state.option.scale) * (diffX / state.option.columnWidth);
    state.ui.currentTime = new Date(state.ui.currentTime.getTime() + diffTime);
  });
};

export const updateViewport = (rect: { width: number; height: number; }) => {
  State.update(state => {
    state.ui.viewportWidth = rect.width;
    state.ui.viewportHeight = rect.height;
  });
};

export const selectTask = (taskId: State.Task.TaskId | undefined) => {
  State.update(state => {
    state.ui.selectedTaskId = taskId;
  });
  setTimeout(() => moveToSelectedTask());
};

export const moveToSelectedTask = () => {
  State.update(state => {
    if (!state.ui.selectedTaskId) return;

    const target = State.Task.getTask(state.tasks, state.ui.selectedTaskId)!;
    const { scale, columnWidth } = state.option;
    const { currentTime, viewportWidth } = state.ui;
    const startTime = currentTime;
    const finishTime = new Date(currentTime.getTime() + (viewportWidth / columnWidth * State.Option.scaleTime(scale)));

    if (startTime.getTime() > target.finishedAt.getTime()) {
      state.ui.currentTime = new Date(target.startedAt.getTime() - State.Option.scaleTime(scale) * 14);
    } else if (target.startedAt.getTime() > finishTime.getTime()) {
      state.ui.currentTime = new Date(target.startedAt.getTime() - State.Option.scaleTime(scale) * 14);
    }
  });
};

