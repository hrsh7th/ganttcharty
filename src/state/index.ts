import { define, Select } from 'srimmer';
import * as UI from './ui';
import * as Task from './task';
import * as Option from './option';

export { Select, Task, Option, UI };

export type State = {
  ui: UI.UI;
  option: Option.Option;
  tasks: Task.Task[];
};

export const {
  Provider,
  Consumer,
  update,
  select,
  get,
  set
} = define<State>();

