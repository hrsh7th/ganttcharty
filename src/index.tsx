import React from 'react';
import ReactDOM from 'react-dom';
import * as State from './state';
import { GanttChart } from './component/app/GanttChart';

type Input = {
  option?: any;
  tasks: (State.Task.Task | { startedAt: string; finishedAt: string })[];
};

export { State };

export function render(id: string, state: Input) {
  const element = document.getElementById(id)!;
  State.setup(defaults(element, state));
  ReactDOM.render(<GanttChart />, element);
}

const defaults = (element: HTMLElement, input: Input): State.State => {
  const state = {} as any;
  state.option = State.Option.defaults(input.option || {});
  state.ui = State.UI.defaults(element, input.option);
  state.tasks = State.Task.defaults(input.tasks || []);
  return state as State.State;
};
