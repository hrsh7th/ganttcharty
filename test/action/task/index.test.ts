import diff from 'snapshot-diff';
import { fixture } from '../../';
import * as State from '../../../src/state';
import * as Action from '../../../src/action';

beforeEach(() => {
  State.set(fixture());
});

let state: State.State;

test('moveSelectedTask', () => {
  Action.UI.selectTask(1);

  state = State.get()!;
  Action.Task.moveSelectedTask(State.Option.DAY);
  expect(diff(state, State.get())).toMatchSnapshot();
  state = State.get()!;
  Action.Task.moveSelectedTask(-State.Option.DAY);
  expect(diff(state, State.get())).toMatchSnapshot();
});

