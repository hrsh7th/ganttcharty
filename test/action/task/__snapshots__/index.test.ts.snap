// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`moveSelectedTask 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -18,14 +18,14 @@
      \\"scale\\": \\"day\\",
    },
    \\"tasks\\": Array [
      Object {
        \\"description\\": \\"\\",
-       \\"finishedAt\\": 2018-10-13T15:00:00.000Z,
+       \\"finishedAt\\": 2018-10-14T15:00:00.000Z,
        \\"id\\": 1,
        \\"name\\": \\"default task name\\",
-       \\"startedAt\\": 2018-10-12T15:00:00.000Z,
+       \\"startedAt\\": 2018-10-13T15:00:00.000Z,
      },
      Object {
        \\"description\\": \\"\\",
        \\"finishedAt\\": 2018-10-13T15:00:00.000Z,
        \\"id\\": 2,"
`;

exports[`moveSelectedTask 2`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -18,14 +18,14 @@
      \\"scale\\": \\"day\\",
    },
    \\"tasks\\": Array [
      Object {
        \\"description\\": \\"\\",
-       \\"finishedAt\\": 2018-10-14T15:00:00.000Z,
+       \\"finishedAt\\": 2018-10-13T15:00:00.000Z,
        \\"id\\": 1,
        \\"name\\": \\"default task name\\",
-       \\"startedAt\\": 2018-10-13T15:00:00.000Z,
+       \\"startedAt\\": 2018-10-12T15:00:00.000Z,
      },
      Object {
        \\"description\\": \\"\\",
        \\"finishedAt\\": 2018-10-13T15:00:00.000Z,
        \\"id\\": 2,"
`;
